crypto/serializer/serializer_meth.c

```diff
@@ -0,0 +1,358 @@
+{
+    OSSL_SERIALIZER *serializer = NULL;
+
+    if ((serializer = OPENSSL_zalloc(sizeof(*serializer))) == NULL
+        || (serializer->lock = CRYPTO_THREAD_lock_new()) == NULL) {
+        OSSL_SERIALIZER_free(serializer);
+        ERR_raise(ERR_LIB_SERIALIZER, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    serializer->refcnt = 1;
+
+    return serializer;
+}
+
+int OSSL_SERIALIZER_up_ref(OSSL_SERIALIZER *serializer)
+{
+    int ref = 0;
+
+    CRYPTO_UP_REF(&serializer->refcnt, &ref, serializer->lock);
+    return 1;
+}
+
+void OSSL_SERIALIZER_free(OSSL_SERIALIZER *serializer)
+{
+    int ref = 0;
+
+    if (serializer == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&serializer->refcnt, &ref, serializer->lock);
+    if (ref > 0)
+        return;
+    ossl_provider_free(serializer->prov);
+    CRYPTO_THREAD_lock_free(serializer->lock);
+    OPENSSL_free(serializer);
+}
+
+const OSSL_PROVIDER *OSSL_SERIALIZER_provider(const OSSL_SERIALIZER *serializer)
+{
+    return serializer->prov;
+}
+
+int OSSL_SERIALIZER_number(const OSSL_SERIALIZER *serializer)
+{
+    return serializer->id;
+}
+
+int OSSL_SERIALIZER_is_a(const OSSL_SERIALIZER *serializer, const char *name)
+{
+    if (serializer->prov != NULL) {
+        OPENSSL_CTX *libctx = ossl_provider_library_context(serializer->prov);
+        OSSL_NAMEMAP *namemap = ossl_namemap_stored(libctx);
+
+        return ossl_namemap_name2num(namemap, name) == serializer->id;
+    }
+    return 0;
+}
+
+struct serializer_do_all_data_st {
+    void (*user_fn)(void *method, void *arg);
+    void *user_arg;
+};
+
+static void serializer_do_one(OSSL_PROVIDER *provider,
+                              const OSSL_ALGORITHM *algo,
+                              int no_store, void *vdata)
+{
+    struct serializer_do_all_data_st *data = vdata;
+    OPENSSL_CTX *libctx = ossl_provider_library_context(provider);
+    OSSL_NAMEMAP *namemap = ossl_namemap_stored(libctx);
+    int id = ossl_namemap_add_names(namemap, 0, algo->algorithm_names,
+                                    NAME_SEPARATOR);
+    void *method = NULL;
+
+    if (id != 0)
+        method =
+            serializer_from_dispatch(id, algo->implementation, provider, NULL);
+
+    if (method != NULL) {
+        data->user_fn(method, data->user_arg);
+        OSSL_SERIALIZER_free(method);
+    }
+}
+
+void OSSL_SERIALIZER_do_all_provided(OPENSSL_CTX *libctx,
+                                     void (*fn)(OSSL_SERIALIZER *serializer,
+                                                void *arg),
+                                     void *arg)
+{
+    struct serializer_do_all_data_st data;
+
+    data.user_fn = (void (*)(void *, void *))fn;
+    data.user_arg = arg;
+    ossl_algorithm_do_all(libctx, OSSL_OP_SERIALIZER, NULL,
+                          serializer_do_one, &data);
+}
+
+void OSSL_SERIALIZER_names_do_all(const OSSL_SERIALIZER *serializer,
+                                  void (*fn)(const char *name, void *data),
```

Comment:
maybe just call one of these data fields something else (like dat) OR typedef the callback and just supply that..

Reply:
Not sure I understand why "dat" is better than "data". Typedeffing could be something, do you have a suitable name?

Reply:
Well if you want to go all patterns, then "vistor_fn" as the function typedef?  Or like the rest of openssl, serializer_cb maybe?

Reply:
This sounds like a subkect for a separate PR